components/tidb_query/src/codec/mysql/binary_literal.rs

```diff
@@ -0,0 +1,466 @@
+            head.extend(trimed.as_bytes());
+            box_try!(hex::decode(head))
+        } else {
+            box_try!(hex::decode(trimed.as_bytes()))
+        };
+        Ok(BinaryLiteral(v))
+    }
+
+    /// Parses bit string.
+    /// The string format can be b'val', B'val' or 0bval, val must be 0 or 1.
+    /// See https://dev.mysql.com/doc/refman/5.7/en/bit-value-literals.html
+    pub fn from_bit_str(s: &str) -> Result<Self> {
+        if s.is_empty() {
+            return Err(box_err!("invalid empty string for parsing bit type"));
+        }
+        let trimed = if s.starts_with('b') || s.starts_with('B') {
+            // format is b'val' or B'val'
+            let trimed = s[1..].trim_start_matches('\'');
+            trimed.trim_end_matches('\'')
+        } else if s.starts_with("0b") {
+            s.trim_start_matches("0b")
+        } else {
+            // here means format is not b'val', B'val' or 0bval.
+            return Err(box_err!("invalid hexadecimal format: {}", s));
+        };
+        if trimed.is_empty() {
+            return Ok(BinaryLiteral(vec![]));
+        }
+
+        let aligned_len = (trimed.len() + 7) & (!7);
+        let mut padding_str = "0".repeat(8);
+        padding_str.push_str(trimed);
+        let start = trimed.len() + 8 - aligned_len;
+        let substr = &padding_str[start..];
+        let len = substr.len() >> 3;
+        let mut buf = Vec::with_capacity(len);
+        let mut i = 0;
+        while i < len {
+            let pos = i << 3;
+            let val = box_try!(usize::from_str_radix(&substr[pos..(pos + 8)], 2));
+            buf.push(val as u8);
+            i += 1;
+        }
+        Ok(BinaryLiteral(buf))
+    }
+
+    /// Returns the bit literal representation for the literal.
+    pub fn to_bit_string(&self, trim_zero: bool) -> String {
+        if self.0.is_empty() {
+            return "b''".to_string();
+        }
+        let mut s = String::with_capacity(self.0.len() * 2 + 1);
+        for b in &self.0 {
+            s.push_str(&format!("{:08b}", *b));
+        }
+        if trim_zero {
+            let trimed = s.trim_start_matches('0');
+            if trimed.is_empty() {
+                s.clear();
+                s.push_str("b'0'");
+                return s;
+            }
+            let trimed_len = s.len() - trimed.len();
+            s.drain(0..trimed_len);
+        }
+        s.insert_str(0, "b'");
+        s.push('\'');
+        s
+    }
+
+    /// Returns the int value for the literal.
+    pub fn to_uint(&self, ctx: &mut EvalContext) -> Result<u64> {
+        to_uint(ctx, &self.0)
+    }
+}
+
+impl ToString for BinaryLiteral {
+    fn to_string(&self) -> String {
+        if self.0.is_empty() {
+            return "".to_string();
+        }
+        format!("0x{}", hex::encode(self.0.as_slice()))
+    }
+}
+
+impl Eq for BinaryLiteral {}
+
+impl PartialEq for BinaryLiteral {
+    fn eq(&self, rhs: &Self) -> bool {
+        self.0 == rhs.0
```

Comment:
Should we trim the leading zeros?

Reply:
This behavior does not define in TiDB (But this trait must be implemented in Rust, I think it's does not matter because this part does not use actually).